certora/specs/AccessManager.spec

```diff
@@ -0,0 +1,588 @@
+│ Functions: canCall                                                                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule canCall(env e) {
+    address caller;
+    address target;
+    bytes4 selector;
+
+    // Get relevant values
+    bool   immediate    = canCall_1(e, caller, target, selector);
+    uint32 delay        = canCall_2(e, caller, target, selector);
+    bool   closed       = isTargetClosed(target);
+    uint64 roleId       = getTargetFunctionRole(target, selector);
+    bool   isMember     = hasRole_1(e, roleId, caller);
+    uint32 currentDelay = hasRole_2(e, roleId, caller);
+
+    // Can only execute without delay in specific cases:
+    // - target not closed
+    // - if self-execution: `executionId` must match
+    // - if third party execution: must be member with no delay
+    assert immediate <=> (
+        !closed &&
+        (
+            (caller == currentContract && executionId() == hashExecutionId(target, selector))
+            ||
+            (caller != currentContract && isMember && currentDelay == 0)
+        )
+    );
+
+    // Can only execute with delay in specific cases:
+    // - target not closed
+    // - third party execution
+    // - caller is a member and has an execution delay
+    assert delay > 0 <=> (
+        !closed &&
+        caller != currentContract &&
+        isMember &&
+        currentDelay > 0
+    );
+
+    // If there is a delay, then it must be the caller's execution delay
+    assert delay > 0 => delay == currentDelay;
+
+    // Immediate execute means no delayed execution
+    assert immediate => delay == 0;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ State transitions: getAccess                                                                                        │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule getAccessChangeWait(uint64 roleId, address account) {
+    env e1;
+    env e2;
+
+    // values before
+    mathint getAccess1Before = getAccess_1(e1, roleId, account);
+    mathint getAccess2Before = getAccess_2(e1, roleId, account);
+    mathint getAccess3Before = getAccess_3(e1, roleId, account);
+    mathint getAccess4Before = getAccess_4(e1, roleId, account);
+
+    // time pass: e1 → e2
+    require e1.block.timestamp <= e2.block.timestamp;
+
+    // values after
+    mathint getAccess1After  = getAccess_1(e2, roleId, account);
+    mathint getAccess2After  = getAccess_2(e2, roleId, account);
+    mathint getAccess3After  = getAccess_3(e2, roleId, account);
+    mathint getAccess4After  = getAccess_4(e2, roleId, account);
+
+    // member "since" cannot change as a consequence of time passing
+    assert getAccess1Before == getAccess1After;
```

Comment:
I would check this as it own invariant, it's a question that came to my mind frequently while working on the tests.

Reply:
That is not an invariant though. The value can change, it just can't change by time just passing.

AFAIK, invariant test that some property doesn't change when a function is called.